### PR TITLE
Introduce ra_proc_macro

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR implemented:

1.  Reading dylib path of proc-macro crate from cargo check , similar to how `OUTDIR` is obtained.
2.  Added a new crate `ra_proc_macro` and implement the foot-work for reading result from external proc-macro expander. 
3. Added a struct `ProcMacroClient` , which will be responsible to the client side communication to the External process.

### CODE REVIEW

**Comment by edwin0cheng** [comment]
cc @kiljacken for the dylib path reading part :)

File: crates/ra_proc_macro/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        _derive_name: &str,
+    ) -> Result<Subtree, ExpandError> {
+        // FIXME: do nothing for now
+        Ok(Subtree::default())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ProcMacro {
+    expander: Arc<Box<dyn ProcMacroExpander>>,
```
Comment (Veetaha) [inline crates/ra_proc_macro/src/lib.rs]: What is the rationale for `Arc<Box<T>>`? Maybe just `Arc<T>` ?

File: crates/ra_proc_macro/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
+//! Client-side Proc-Macro crate
+//!
+//! We separate proc-macro expanding logic to an extern program to allow
+//! different implementations (e.g. wasm or dylib loading). And this crate
+//! is used for provide basic infra-structure for commnicate between two
+//! process: Client (RA itself), Server (the external program)
```
[the remark this answers is not in the document]
Comment (edwin0cheng) [reply crates/ra_proc_macro/src/lib.rs]: Fixed.

File: crates/ra_proc_macro/src/lib.rs (hunk 1 above)
Comment (edwin0cheng) [reply crates/ra_proc_macro/src/lib.rs]: Nice catch. I just forget to change it back. thanks!

File: crates/ra_proc_macro/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+version = "0.1.0"
+authors = ["rust-analyzer developers"]
+publish = false
+
+[lib]
+doctest = false
+
+[dependencies]
+ra_tt = { path = "../ra_tt" }
+ra_mbe = { path = "../ra_mbe" }
```
Comment (matklad) [inline crates/ra_proc_macro/Cargo.toml]: Why does this depend on `ra_mbe`? Can we reduce deps here only to tt?

File: crates/ra_db/Cargo.toml [hunk 4]
```diff
@@ -15,4 +15,5 @@ rustc-hash = "1.1.0"
 ra_syntax = { path = "../ra_syntax" }
 ra_cfg = { path = "../ra_cfg" }
 ra_prof = { path = "../ra_prof" }
+ra_proc_macro = { path = "../ra_proc_macro" }
```
Comment (matklad) [inline crates/ra_db/Cargo.toml]: Hm, I wonder if we can minimize the interface here...

Can we use something like this instead?

```
struct CrateData {
   proc_macros: FxHashMap<SmolStr, Arc<dyn Fn(tt::TokenTree) -> tt::TokenTree>>
}
```

Ie, define a really minimal proc macro interface directly in the `ra_db` crate, where the minimal interface is ideally an `dyn Fn()`, or maybe a named trait.

**Comment by matklad** [comment]
Also, thanks @edwin0cheng for splitting changes which are small diff-wise, but are large architecturally, into separate PRs! This is massively helpful!

**Comment by edwin0cheng** [comment]
> Ideally, the ra_db should not depend only any proc-macro specfic, and just export a trait. It should operate on tt::TokenStream, so a dep on ra_tt is required.

In other words, if we export a trait in `ra_db`, Could `ra_proc_macro`depends on `ra_db` ? Or we defined a wrapper in `ra_project_model` to bridge them ?

**Comment by matklad** [comment]
Good question! Yeah, I think it would also be good if `ra_proc_macro` didn't depned on `ra_db`... 

I think bridging them in project-model might work, but, if the expansion trait is really minimal and only depends on tt, we can just put it in the `ra_tt`?

**Comment by edwin0cheng** [comment]
I think it is plausible, I could imagine it looks like following: 

```rust
enum ExpandError {
...
}

trait TokenExpander: Debug + Send + Sync + RefUnwindSafe {
    fn exander(&self, subtree: &Subtree, attrs: Option<&Subtree>) -> Result<Subtree, ExpandError>;
}
```

The trait bound here is need in `CrateData`, as It will be stored in salsa.

**Comment by kiljacken** [comment]
Cargo interacting code looks reasonable to me 👍 

Good job, this is some really cool work!

File: crates/ra_db/src/input.rs [hunk 5]
```diff
@@ -115,6 +116,9 @@ impl Display for CrateName {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct ProcMacroId(pub usize);
```
[the remark this answers is not in the document]
Comment (edwin0cheng) [reply crates/ra_db/src/input.rs]: Fixed

File: crates/ra_db/Cargo.toml (hunk 4 above)
Comment (edwin0cheng) [reply crates/ra_db/Cargo.toml]: I changed to :
```rust

#[derive(Debug, Clone)]
pub struct ProcMacro {
    pub name: SmolStr,
    pub expander: Arc<dyn TokenExpander>,
}

impl Eq for ProcMacro {}
impl PartialEq for ProcMacro { ... }

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CrateData {
   ...
    pub proc_macro: Vec<ProcMacro>,
}
```

**Comment by edwin0cheng** [comment]
@matklad, I removed the dependency for mbe and refactoring a bit for salsa trait constraints. You could review it again. 👍

File: crates/ra_proc_macro/Cargo.toml (hunk 3 above)
Comment (edwin0cheng) [reply crates/ra_proc_macro/Cargo.toml]: Fixed.

**Pull request merged**